aten/src/ATen/core/CachingHostAllocator.h

```diff
@@ -220,50 +224,101 @@ struct alignas(hardware_destructive_interference_size) HostStatsStaged {
  * we hold the same locks as empty_cache, to ensure the fidelity of the stats.
  */
 
+// Generic per-pool structures for the host caching allocator. These are
+// templated so they can reference the allocator's template parameters B and E.
+template <typename S_, typename E_, typename B_>
+struct HostBlockPool {
+  HostBlockPool() = default;
+  HostBlockPool(const HostBlockPool&) = delete;
+  HostBlockPool& operator=(const HostBlockPool&) = delete;
+
+  alignas(hardware_destructive_interference_size) std::mutex blocks_mutex_;
+  ska::flat_hash_set<B_*> blocks_; // all blocks in this pool
+  ska::flat_hash_map<void*, B_*> ptr_to_block_;
+
+  // Per-size free lists guarded by their own mutexes.
+  alignas(hardware_destructive_interference_size) std::vector<FreeBlockList<B_>> free_list_ =
+      std::vector<FreeBlockList<B_>>(MAX_SIZE_INDEX);
+
+  // Events pending for blocks in this pool.
+  alignas(hardware_destructive_interference_size) std::mutex events_mutex_;
+  std::deque<std::pair<E_, B_*>> events_;
+};
+
+// The HostBlockPool owned by one HostPrivatePool cannot share blocks
+// with any other HostPrivatePool. This invariant is necessary to
+// implement cuda graph capture correctly, which depends upon virtual
+// addresses staying alive for the duration of a graph's existence.
+template <typename S_, typename E_, typename B_>
```

Comment:
Is there an equivalent concept for the `CUDACachingAllocator`? If not, why didn't we need a private BlockPool there?

Reply:
There is a similar concept in CUDACachingAllocator